contracts/finality/src/state/allowlist.rs

```diff
@@ -1,49 +1,163 @@
 use cosmwasm_std::Storage;
-use cw_storage_plus::Map;
+use cw_storage_plus::{SnapshotItem, Strategy};
+use std::collections::HashSet;
 
 use crate::error::ContractError;
 use hex;
 
-/// Map of allowed finality provider BTC public keys (as bytes) to a unit value
-pub(crate) const ALLOWED_FINALITY_PROVIDERS: Map<&[u8], ()> =
-    Map::new("allowed_finality_providers");
+// Storage key constants
+const ALLOWED_FINALITY_PROVIDERS_KEY: &str = "allowed_finality_providers";
+const ALLOWED_FINALITY_PROVIDERS_CHECKPOINTS_KEY: &str = "allowed_finality_providers__checkpoints";
+const ALLOWED_FINALITY_PROVIDERS_CHANGELOG_KEY: &str = "allowed_finality_providers__changelog";
 
-/// Check if a finality provider is in the allowlist
+/// SnapshotItem of allowed finality provider BTC public keys stored as a HashSet
+pub(crate) const ALLOWED_FINALITY_PROVIDERS: SnapshotItem<HashSet<Vec<u8>>> = SnapshotItem::new(
+    ALLOWED_FINALITY_PROVIDERS_KEY,
+    ALLOWED_FINALITY_PROVIDERS_CHECKPOINTS_KEY,
+    ALLOWED_FINALITY_PROVIDERS_CHANGELOG_KEY,
+    Strategy::EveryBlock,
+);
+
+/// Check if a finality provider is in the allowlist (at current Babylon height)
 pub fn ensure_fp_in_allowlist(
     storage: &dyn Storage,
     fp_btc_pk_bytes: &[u8],
 ) -> Result<(), ContractError> {
-    ALLOWED_FINALITY_PROVIDERS
-        .has(storage, fp_btc_pk_bytes)
-        .then_some(())
-        .ok_or(ContractError::FinalityProviderNotAllowed(hex::encode(
+    let fp_set = ALLOWED_FINALITY_PROVIDERS
+        .may_load(storage)?
+        .unwrap_or_default();
+
+    if fp_set.contains(fp_btc_pk_bytes) {
+        Ok(())
+    } else {
+        Err(ContractError::FinalityProviderNotAllowed(hex::encode(
             fp_btc_pk_bytes,
         )))
+    }
 }
 
-/// Add a finality provider to the allowlist
-pub fn add_finality_provider_to_allowlist(
+/// Add finality providers to the allowlist at a specific Babylon height
+pub fn add_finality_providers_to_allowlist(
     storage: &mut dyn Storage,
-    fp_btc_pk_bytes: &[u8],
+    fp_btc_pk_bytes_list: &[Vec<u8>],
```

Comment:
Not sure if we can do

```suggestion
    fp_btc_pk_bytes_list: &[&[u8]],
```